node/overseer/src/lib.rs

```diff
@@ -919,15 +953,15 @@ where
 	async fn block_imported(&mut self, block: BlockInfo) -> SubsystemResult<()> {
 		let mut update = ActiveLeavesUpdate::default();
 
-		if let Some(parent) = block.number.checked_sub(1).and_then(|number| self.active_leaves.take(&(block.parent_hash, number))) {
-			update.deactivated.push(parent.0);
+		if let Some(_number) = self.active_leaves.remove(&block.parent_hash) {
```

Comment:
Yes, I've changed the logic.
> I don't know why active_leaves was defined as a HashMap<Hash, BlockNumber> and if there are any downsides to making it a HashMap.

the question is: is it possible to get two `BlockInfo`'s with the same hash, but different block numbers? <- that seems like a bug to me

initial logic was implemented in https://github.com/paritytech/polkadot/pull/1188 cc @montekki

Reply:
> the question is: is it possible to get two BlockInfo's with the same hash, but different block numbers? <- that seems like a bug to me

I think no, my reasoning is the following:
* the client is the one who will call `block_imported` and provide `BlockInfo`
* the client has a db, with a 1-to-1 mapping hash -> number (https://github.com/paritytech/substrate/blob/dbf2163250833e6ac898e7f6c3c8f89f08a7c19d/primitives/blockchain/src/backend.rs#L41)
* so will never receive a `BlockInfo` with the same hash, but different numbers
* instead of having block number in `BlockInfo`, we could request it via `ChainApi` on `block_finalized` where we actually need it
* but since it's the client who provided us `BlockInfo` in the first place, we could avoid unnecessary requests to the db

Please correct me if I'm wrong.

Reply:
Ok, your logic makes sense. I am persuaded and withdraw my objection to this logic change.

Reply:
can add some defensive assertions
```patch
diff --git a/node/overseer/src/lib.rs b/node/overseer/src/lib.rs
index 09aca23b..566114e2 100644
--- a/node/overseer/src/lib.rs
+++ b/node/overseer/src/lib.rs
@@ -59,7 +59,7 @@ use std::pin::Pin;
 use std::sync::Arc;
 use std::task::Poll;
 use std::time::Duration;
-use std::collections::HashMap;
+use std::collections::{hash_map, HashMap};
 
 use futures::channel::{mpsc, oneshot};
 use futures::{
@@ -934,16 +934,24 @@ where
 	async fn block_imported(&mut self, block: BlockInfo) -> SubsystemResult<()> {
 		let mut update = ActiveLeavesUpdate::default();
 
-		if let Some(_number) = self.active_leaves.remove(&block.parent_hash) {
+		if let Some(number) = self.active_leaves.remove(&block.parent_hash) {
+			if let Some(expected_parent_number) = block.number.checked_sub(1) {
+				debug_assert_eq!(expected_parent_number, number);
+			}
 			update.deactivated.push(block.parent_hash);
 			self.on_head_deactivated(&block.parent_hash);
 		}
 
-		if self.active_leaves.get(&block.hash).is_none() {
-			update.activated.push(block.hash);
-			self.active_leaves.insert(block.hash, block.number);
-			self.on_head_activated(&block.hash);
-		}
+		match self.active_leaves.entry(block.hash) {
+			hash_map::Entry::Vacant(entry) => {
+				update.activated.push(block.hash);
+				entry.insert(block.hash, block.number);
+				self.on_head_activated(&block.hash);
+			},
+			hashMap::Entry::Occupied(entry) => {
+				debug_assert_eq!(entry.get(), block.number);
+			}
+	}
 
 		self.clean_up_external_listeners();
```
but it clutters the code as a downside

Reply:
I quite like that, despite the clutter.